libraries/botbuilder-dialogs/tests/attachmentPrompt.test.js

```diff
@@ -0,0 +1,180 @@
+            }
+        }));
+        
+        adapter.send('Hello')
+        .assertReply('Please send an attachment.')
+        .send(invalidMessage)
+        .assertReply('Please try again.')
+        .send(answerMessage)
+        .assertReply('test1')
+        done();
+    });
+
+    it('should send ignore retryPrompt if validator replies.', function (done) {
+        const adapter = new TestAdapter(async (turnContext) => {
+            const dc = await dialogs.createContext(turnContext);
+
+            const results = await dc.continue();
+            if (!turnContext.responded && !results.hasActive && !results.hasResult) {
+                await dc.prompt('prompt', { prompt: 'Please send an attachment.', retryPrompt: 'Please try again.' });
+            } else if (!results.hasActive && results.hasResult) {
+                assert(Array.isArray(results.result) && results.result.length > 0);
+                const attachment = results.result[0];
+                await turnContext.sendActivity(`${attachment.content}`);
+            }
+        });
+
+        const convoState = new ConversationState(new MemoryStorage());
+        adapter.use(convoState);
+
+        const dialogState = convoState.createProperty('dialogState');
+        const dialogs = new DialogSet(dialogState);
+        dialogs.add(new AttachmentPrompt('prompt', async (context, prompt) => {
+            assert(context);
+            assert(prompt);
+            
+            if (!prompt.recognized.value) {
+                await context.sendActivity('Bad input.');
+            } else {
+                prompt.end(prompt.recognized.value);
+            }
+        }));
+        
+        adapter.send('Hello')
+        .assertReply('Please send an attachment.')
+        .send(invalidMessage)
+        .assertReply('Bad input.')
+        .send(answerMessage)
+        .assertReply('test1')
+        done();
+    });
+
+    it('should not send any retryPrompt if no prompt is specified.', function (done) {
+        const adapter = new TestAdapter(async (turnContext) => {
+            const dc = await dialogs.createContext(turnContext);
+
+            const results = await dc.continue();
+            if (!turnContext.responded && !results.hasActive && !results.hasResult) {
+                await dc.begin('prompt');
+            } else if (!results.hasActive && results.hasResult) {
+                assert(Array.isArray(results.result) && results.result.length > 0);
+                const attachment = results.result[0];
+                await turnContext.sendActivity(`${attachment.content}`);
+            }
+        });
+
+        const convoState = new ConversationState(new MemoryStorage());
+        adapter.use(convoState);
+
+        const dialogState = convoState.createProperty('dialogState');
+        const dialogs = new DialogSet(dialogState);
+        dialogs.add(new AttachmentPrompt('prompt'));
+        
+        adapter.send('Hello')
```

Comment:
There's a difference in how this test operates versus the ConfirmPrompt tests because of how the ConfirmPrompt calls the ChoiceFactory through the base Prompt class (via [Prompt.appendChoices()][appendChoices]). @stevenic do you have a recommendation for how to get the two to act similarly?


  [appendChoices]: https://github.com/Microsoft/botbuilder-js/blob/master/libraries/botbuilder-dialogs/src/prompts/prompt.ts#L164